// @adonisjs/shield/build/standalone.js
"use strict";
/*
 * @adonisjs/shield
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.dnsPrefetchFactory = exports.frameGuardFactory = exports.noSniffFactory = exports.hstsFactory = exports.csrfFactory = exports.cspFactory = void 0;
var csp_1 = require("./src/csp");
Object.defineProperty(exports, "cspFactory", { enumerable: true, get: function () { return csp_1.cspFactory; } });
var csrf_1 = require("./src/csrf");
Object.defineProperty(exports, "csrfFactory", { enumerable: true, get: function () { return csrf_1.csrfFactory; } });
var hsts_1 = require("./src/hsts");
Object.defineProperty(exports, "hstsFactory", { enumerable: true, get: function () { return hsts_1.hstsFactory; } });
var noSniff_1 = require("./src/noSniff");
Object.defineProperty(exports, "noSniffFactory", { enumerable: true, get: function () { return noSniff_1.noSniffFactory; } });
var frameGuard_1 = require("./src/frameGuard");
Object.defineProperty(exports, "frameGuardFactory", { enumerable: true, get: function () { return frameGuard_1.frameGuardFactory; } });
var dnsPrefetch_1 = require("./src/dnsPrefetch");
Object.defineProperty(exports, "dnsPrefetchFactory", { enumerable: true, get: function () { return dnsPrefetch_1.dnsPrefetchFactory; } });
