// @adonisjs/shield/build/src/ShieldMiddleware/index.js
"use strict";
/*
 * @adonisjs/shield
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShieldMiddleware = void 0;
const standalone_1 = require("@adonisjs/core/build/standalone");
const shield = __importStar(require("../../standalone"));
/**
 * Shield middleware to protect web applications against common
 * web attacks
 */
let ShieldMiddleware = exports.ShieldMiddleware = class ShieldMiddleware {
    constructor(application) {
        this.application = application;
        this.config = this.application.container.resolveBinding('Adonis/Core/Config').get('shield', {});
        this.encryption = this.application.container.resolveBinding('Adonis/Core/Encryption');
        /**
         * Resolve view provider, when exists
         */
        this.view = this.application.container.hasBinding('Adonis/Core/View')
            ? this.application.container.resolveBinding('Adonis/Core/View')
            : undefined;
        /**
         * Actions to be performed
         */
        this.actions = [
            shield.csrfFactory(this.config.csrf || {}, this.encryption, this.view),
            shield.cspFactory(this.config.csp || {}),
            shield.dnsPrefetchFactory(this.config.dnsPrefetch || {}),
            shield.frameGuardFactory(this.config.xFrame || {}),
            shield.hstsFactory(this.config.hsts || {}),
            shield.noSniffFactory(this.config.contentTypeSniffing || {}),
        ];
    }
    /**
     * Handle request
     */
    async handle(ctx, next) {
        for (let action of this.actions) {
            await action(ctx);
        }
        await next();
    }
};
exports.ShieldMiddleware = ShieldMiddleware = __decorate([
    (0, standalone_1.inject)(['Adonis/Core/Application']),
    __metadata("design:paramtypes", [Object])
], ShieldMiddleware);
